kernel/src/scan/log_replay.rs

```diff
@@ -103,7 +109,34 @@ impl AddRemoveDedupVisitor<'_> {
 
         // Process both adds and removes, but only return not already-seen adds
         let file_key = FileActionKey::new(path, dv_unique_id);
-        Ok(!self.check_and_record_seen(file_key) && is_add)
+        let have_seen = self.check_and_record_seen(file_key);
+        if is_add && !have_seen {
+            // compute transform here
+            if let Some(ref transform) = self.transform {
```

Comment:
Nit: I like to avoid nesting where possible. I wonder if we can do early returns or factor this out into a `resolve_transform_expr` function.